sql: introduce MultiRowFetcher for one-pass row fetches for multiple (interleaved) tables

richardwu opened the pull request:
This PR introduces `MultiRowFetcher` which generalizees `RowFetcher` to
respond to and return rows from multiple tables from the same set of
spans.

`MultiRowFetcher` is initialized with a slice of `MultiRowFetcherArgs`
which encapsulates the arguments for one table in `RowFetcher`.

One key point is that the `reverse bool` argument that exists in
`RowFetcher` can only be enabled on a per-fetcher basis: one cannot scan
in arbitrary directions on a per-table basis (since row kvfetcher works
by scanning KV pairs sequentially).

`MultiRowFetcher` is necessary in order to fetch interleaved tables
efficiently in one-pass. See the RFC for interleaved tables here
https://github.com/cockroachdb/cockroach/pull/19028.

There are no tests yet since I wanted to get some comments on this approach. `MultiRowFetcher`
is backwards compatible with `RowFetcher`: only the `Init` parameters and the response from `NextRow/NextRowDecoded` differs slightly.

cc: @vivekmenezes

richardwu committed:
sql: introduce MultiRowFetcher for one-pass row fetches for multiple
(interleaved) tables

This PR introduces `MultiRowFetcher` which generalizees `RowFetcher` to
respond to and return rows from multiple tables from the same set of
spans.

`MultiRowFetcher` is initialized with a slice of `MultiRowFetcherArgs`
which encapsulates the arguments for one table in `RowFetcher`.

One key point is that the `reverse bool` argument that exists in
`RowFetcher` can only be enabled on a per-fetcher basis: one cannot scan
in arbitrary directions on a per-table basis (since row kvfetcher works
by scanning KV pairs sequentially).

`MultiRowFetcher` is necessary in order to fetch interleaved tables
efficiently in one-pass. See the RFC for interleaved tables here
https://github.com/cockroachdb/cockroach/pull/19028.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19228)
<!-- Reviewable:end -->

richardwu commented:
So I've been griping about how one would check for which table a given decoded key belongs to. The above implementation currently iterates through each table passed into `MultIRowFetcher` and tries to decode it for that table (extracts `TableID` and `IndexID` and checks for equivalence).

This iteration happens for every key, which may be especially bad if there are many tables (more than a couple) passed into `MultiRowFetcher`). One idea I thought of is we pre-process each table's row prefix (including all ancestor prefixes) during `Init`. We then extract the entire key prefix (with only its ancestors' table IDs, index IDs, and interleave sentinels) once for each key (this logic would be similar to [this part of `DecodeIndexKey`] (https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/table.go#L749#L771)). We then lookup the table with said prefix, and continue decoding the rest of the `KeyVals`. 

In fact, I wrote something similar in a `[table.go] DecodeMultiIndexKey` function which would actually shortcut to the terminating ancestor in [this part of `DecodeIndexKey`](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sqlbase/table.go#L754#L756): it would do a lookup for the table with the `ancestor.TableID` and `ancestor.IndexID` via a map). This accomplishes the shortcutting as described above but re-uses most of the code from `DecodeIndexKey`.

Another optimization point is the ordering of the tables in the `tables []MultiRowFetcherArgs` parameter of `MultiRowFetcher.Init`. That is given `tables = {<tableA>, <tableB>}` and `<tableB>` has significantly more keys in the scan than `<tableA>`, we will more often successfully decode a key on the 2nd iteration. On the contrary, if we ordered `tables = {<tableB>, <tableA>}`, then we successfully decode a key on the 1st iteration more often. Intuitively, children rows are often orders of magnitude more numerous than their parent rows, so this scenario is relevant where `<tableA> = parent` and `<tableB> = child`. Furthermore, we expect keys for a given row or even for a `child` table to be sequentially scanned, so perhaps it's better to cache the index to start the iteration of `tables` after each successful key decoding (e.g. if `<tableB>` key is decoded, we start the next key decoding with `<tableB>`). 

Let me know your thoughts on the current solution, bearing in mind that `len(tables) = 2` for this first implementation of interleaved tables.

RaduBerinde commented:
Hi Richard,

Sorry I didn't get to look at this yet, I will try to do it soon. In the meantime I'm CCing @danhhz who worked on the original interleaving code and encodings, he may be best to answer your questions above (I will think about it as well).

RaduBerinde commented:
Yes, checking the prefix is what I would go with. Technically all we need to do is to check against the longest prefix and figure out which table it is based on how much of it matched (we can precalculate how much of the prefix matches for each table).

E.g. if we have a hierarchy of three tables:

```
/table1/<k1>/<interleavedTableSentinel>/k2/2/<interleavedTableSentinel>/k3/
            ^first table matches to here    ^second table matches to here
```

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-Kw_XUYy5QOeXLwFRPgd:b3fqpq0)*
<!-- Sent from Reviewable.io -->

richardwu commented:
I introduced the concept of "equivalence prefixes", which basically defines the equivalence class for a table by its entire prefix.

For example, the hierarchy
```
parent
  child
     grandchild
```
will produce the following equivalence prefixes for each table
```
parent:          /<parent table id>/<parent index id>
child:             /<parent table id>/<parent index id>/#/<child table id>/<child index id> 
grandchild:   /<parent table id>/<parent index id>/#/<child table id>/<child index id>/#/<grandchild table id>/<grandchild index id>
```

Some an index from a `child` KV pair which might look something like
```
/50/1/20/#/51/1/5/0
```
with `parent_id = 20` and `id=5` will produce the equivalence prefix
```
/50/1/#/51/1
```

The lookup for the table with the equivalent prefix is currently done through slice iteration (this is more efficient than a hash map for `n=2`). We can later refactor this to hash on the equivalence prefix and use a `map` instead once we use this for a significant # of tables.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-KwaxrhUvCjUvJyjkCg9:b-fy1770)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
The equivalence prefixes stuff makes sense to me. I haven't yet looked over the multi-row-fetcher code.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/table.go, line 2213 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweRCjOBE9ojxfED6X1:-KweRCjOBE9ojxfED6X2:b-chsyw9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2213)):*
> ```Go
> //	<child@secondary>
> // are represented as (in normal encoding format)
> //    <parent@primary>: /<parent table id>/<parent's primary index id>
> ```

This should explain that we are stripping out the values at each level. Maybe give an example that from the key
`/<parent table id>/<parent index id>/<val1>/<val2>/#/<child table id>/<child index id>/<val3>/<val4>`
we extract the prefix
`/<parent table id>/<parent index id>/#/<child table id>/<child index id>`

Also this is not quite a prefix, maybe we need a different name. Maybe "fingerprint"?

---

*[pkg/sql/sqlbase/table.go, line 2217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweRp8GiZh90Ny_L7v_:-KweRp8GiZh90Ny_L7va:b-wlu374) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2217)):*
> ```Go
> // Equivalence prefixes allow us to associate an index key with its table
> // without having to invoke DecodeIndexKey multiple times.
> func ExtractIndexKeyEquivPrefix(key []byte) (prefix []byte, err error) {
> ```

Ideally we wouldn't allocate stuff for each row; the multi-row-fetcher could pass in a buffer that can be reused

---

*[pkg/sql/sqlbase/table.go, line 2257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweSBcO2eUz2-PpFKC2:-KweSBcO2eUz2-PpFKC3:bcn2nz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2257)):*
> ```Go
> 			// We insert an interleave sentinel and continue extracting the
> 			// descendant's IDs.
> 			if _, err := out.Write(encoding.EncodeNotNullDescending(scratch[:0])); err != nil {
> ```

Why aren't we encoding directly into the key? We should use a `[]byte` instead of a `bytes.Buffer` and do `key = EncodeNotNullDesc(key)`

Also, NYC but it would be nice if we had a wrapper `sql.EncodeInterleaveSentinel` around `encoding.EncodeNotNullDescending` (the call is confusing when you read the code).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-KweTGm20Ry_66D3NjCR:b84929a)*
<!-- Sent from Reviewable.io -->

richardwu commented:
PTAL, cleaned up the `bytes.Buffer` stuff and scratch space.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/sqlbase/table.go, line 2213 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweRCjOBE9ojxfED6X1:-KwgizDGuXvmSsKEe4D5:brayhit) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2213)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This should explain that we are stripping out the values at each level. Maybe give an example that from the key
`/<parent table id>/<parent index id>/<val1>/<val2>/#/<child table id>/<child index id>/<val3>/<val4>`
we extract the prefix
`/<parent table id>/<parent index id>/#/<child table id>/<child index id>`

Also this is not quite a prefix, maybe we need a different name. Maybe "fingerprint"?
</blockquote></details>

Updated the function comment and I renamed it to "signature" if that's alright.

---

*[pkg/sql/sqlbase/table.go, line 2217 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweRp8GiZh90Ny_L7v_:-KwgXuD5Di9YdHilmY_1:bxnmy90) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2217)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ideally we wouldn't allocate stuff for each row; the multi-row-fetcher could pass in a buffer that can be reused
</blockquote></details>

Oops, meant to refactor this. Scratch space isn't necessary now that we're not using `bytes.Buffer`. Good catch!

---

*[pkg/sql/sqlbase/table.go, line 2257 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KweSBcO2eUz2-PpFKC2:-KwgmDl_ioMp8EORrLql:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bf5bea44a04efc02a4852dd0058d612a27593c5/pkg/sql/sqlbase/table.go#L2257)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why aren't we encoding directly into the key? We should use a `[]byte` instead of a `bytes.Buffer` and do `key = EncodeNotNullDesc(key)`

Also, NYC but it would be nice if we had a wrapper `sql.EncodeInterleaveSentinel` around `encoding.EncodeNotNullDescending` (the call is confusing when you read the code).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-Kwguaeopu1MtLYvWg72:b-odsmfi)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/sqlbase/table.go, line 2250 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kwh-hc6I6ot8YPc5eRT:-Kwh-hc79W7OLUlXh0Ky:b9syosh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b9c060a96fb92f2010727e60d61b380e6b06816/pkg/sql/sqlbase/table.go#L2250)):*
> ```Go
> 			}
> 			// Append column values within index key to return.
> 			colVals = append(colVals, key[:len]...)
> ```

I realized that we can in fact memoize all the column values we come across instead of having to iterate through `DecodeIndexKey` again. All bytes we come across are appended to a buffer that will be returned.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/table.go, line 2255 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kwkp9j2zH_Pogtigw13:-Kwkp9j2zH_Pogtigw14:b56z9vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/table.go#L2255)):*
> ```Go
> 			// The current ancestor is not one of the specified ancestors.
> 			// We can short-circuit and return false.
> 			if validAncestorSignatures != nil && !validAncestorSignatures[string(signature)] {
> ```

I added another optimization when the signature for an index key is being decoded: if passed a set of ancestor signatures, it will short-circuit if the current ancestor signature __of the index key__ isn't in the set.

This allows us to practically short-circuit interleaved keys that are a deeper hierarchy level than the tables we're scanning for. For example
```
/parent
   /parent/#/child
      /parent/#/child/#/grandchild/#/.../(grand)*child
   /parent/#/child
```
If we are scanning for `parent` and `child` rows and we are checking for the signature of the index key for `grandgrandchild`, it will short-circuit at the `grandchild` prefix instead of continuing until it reaches the end of the key.

For many keys, this should help out a bit and actually aligns us with how [`DecodeIndexKey` does it](https://github.com/cockroachdb/cockroach/blob/542195d9c1ed2b4b506a0039b1b95cdb1bb4ab1a/pkg/sql/sqlbase/table.go#L763). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/multirowfetcher.go, line 431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8Y5Y8hampW9sZWVdU:-Kx8Y5Y8hampW9sZWVdV:bbgc1sr) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/multirowfetcher.go#L431)):*
> ```Go
> 	// TODO(richardwu): Once MultiRowFetcher is used for more than two
> 	// tables, this logic would be cleaner and perhaps more efficient if
> 	// a map is used instead of a slice for mrf.tables.
> ```

We are already looking it up in a map `allEquivSignatures`.. Why look it up again? I think we should just have that as a `map[string]int` where the int is the table id (`0` to `len(mrf.tables)-1`) and have `IndexKeyEquivSignature` return the table id.

---

*[pkg/sql/sqlbase/table.go, line 2220 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8X55ktalmeEknVu-2:-Kx8X55ktalmeEknVu-3:b-er8udo) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/table.go#L2220)):*
> ```Go
> // Equivalence signatures allow us to associate an index key with its table
> // without having to invoke DecodeIndexKey multiple times.
> // If validAncestorSignatures is not nil, then IndexKeyEquivSignature will
> ```

What's the use case for `validAncestorSignatures=nil`? Just testing?

---

*[pkg/sql/sqlbase/table.go, line 2235 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8WXTJbcR_fqyACj5m:-Kx8WXTKy3esQ0szdtIo:brgu537) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/table.go#L2235)):*
> ```Go
> ) (signature []byte, rest []byte, success bool, err error) {
> 	var isSentinel bool
> 	signature = make([]byte, 0, len(key))
> ```

We should try to avoid per-row allocations as much as possible. I think we can reuse the same buffers across rows. (Can be in a separate PR though if you prefer)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 5 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/sqlbase/multirowfetcher.go, line 431 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8Y5Y8hampW9sZWVdU:-Kx97PzJM82KuE3HZaJV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/multirowfetcher.go#L431)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We are already looking it up in a map `allEquivSignatures`.. Why look it up again? I think we should just have that as a `map[string]int` where the int is the table id (`0` to `len(mrf.tables)-1`) and have `IndexKeyEquivSignature` return the table id.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 2220 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8X55ktalmeEknVu-2:-Kx9Kl8X8lc7s3Dn013x:b-26rh17) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/table.go#L2220)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What's the use case for `validAncestorSignatures=nil`? Just testing?
</blockquote></details>

Yup it was originally for testing. In hind sight, this is not good practice and I've updated the tests to pass a valid `validEquivSignatures` (renamed it from `validAncestorSignatures` since it includes the key's table's signature too) map.

---

*[pkg/sql/sqlbase/table.go, line 2235 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-Kx8WXTJbcR_fqyACj5m:-Kx9L06AZeNxh62b6NpX:b-7ibr6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/350d36b3987ff7e559130a10ab96fb1405e96938/pkg/sql/sqlbase/table.go#L2235)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should try to avoid per-row allocations as much as possible. I think we can reuse the same buffers across rows. (Can be in a separate PR though if you prefer)
</blockquote></details>

Refactored this so `MultiRowFetcher` passes down a buffer. Good catch! 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
:100: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>   Great work!

---

Reviewed 1 of 4 files at r3, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/sqlbase/multirowfetcher.go, line 476 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxA8_ePDb9Ujvzw6a2O:-KxA8_ePDb9Ujvzw6a2P:b-d7ca8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b352b91c98bfe65ca145ced8e27aed5a946d6ad/pkg/sql/sqlbase/multirowfetcher.go#L476)):*
> ```Go
> 	}
> 
> 	tableIdx, found := mrf.allEquivSignatures[string(keyEquivSignature)]
> ```

this could be returned by IndeKeyEquivSignature so we don't look it up again.

---

*[pkg/sql/sqlbase/table.go, line 2256 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxA82kMoVIZBfus9ki8:-KxA82kMoVIZBfus9ki9:bjbltlj) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b352b91c98bfe65ca145ced8e27aed5a946d6ad/pkg/sql/sqlbase/table.go#L2256)):*
> ```Go
> func IndexKeyEquivSignature(
> 	key []byte, validEquivSignatures map[string]int, signature []byte, rest []byte,
> ) ([]byte, []byte, bool, error) {
> ```

[nit] these should have names (or at least documented in the comment)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-KxA96s5MjryHyuCX7DS:ba8tqvf)*
<!-- Sent from Reviewable.io -->

richardwu commented:
I'm going to work on a few "unit" tests for `MultiRowFetcher` and I'll ping with the updated revision.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/multirowfetcher.go, line 476 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxA8_ePDb9Ujvzw6a2O:-KxANLp4N6ZCcp5HT0xq:b-fdpa7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b352b91c98bfe65ca145ced8e27aed5a946d6ad/pkg/sql/sqlbase/multirowfetcher.go#L476)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this could be returned by IndeKeyEquivSignature so we don't look it up again.
</blockquote></details>

Done. Now that we're returning the table index directly, it begs the question whether it's worthwhile to return the signature of the key itself (it's not being used in `MultiRowFetcher`). For completeness as an exported method in `sqlbase/table.go`, it would make sense for it to also return the signature. A criticism against this is only `MultiRowFetcher` calls this function, so maybe we should just move the `EquivSignature` methods to `MultiRowFetcher` (and un-export them).

Thoughts?

---

*[pkg/sql/sqlbase/table.go, line 2256 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxA82kMoVIZBfus9ki8:-KxANMqCKWBsO_x_r-Hs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b352b91c98bfe65ca145ced8e27aed5a946d6ad/pkg/sql/sqlbase/table.go#L2256)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] these should have names (or at least documented in the comment)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-KxAO2hOUk6N6H4zDnV7:b1p5mk8)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/sql/sqlbase/multirowfetcher.go, line 476 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxA8_ePDb9Ujvzw6a2O:-KxDUTcrB06qnM5fcPbJ:b-qyr78z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2b352b91c98bfe65ca145ced8e27aed5a946d6ad/pkg/sql/sqlbase/multirowfetcher.go#L476)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

Done. Now that we're returning the table index directly, it begs the question whether it's worthwhile to return the signature of the key itself (it's not being used in `MultiRowFetcher`). For completeness as an exported method in `sqlbase/table.go`, it would make sense for it to also return the signature. A criticism against this is only `MultiRowFetcher` calls this function, so maybe we should just move the `EquivSignature` methods to `MultiRowFetcher` (and un-export them).

Thoughts?
</blockquote></details>

The way it works now, the function only returns something useful if the signature is in the map, so returning the signature is not very useful (the caller knows the signature based on the ID). I would keep it in `sqlbase`, we can always extend/generalize it later if we ever need to.

---

*[pkg/sql/sqlbase/table.go, line 2262 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxDTmXRKsV-G1M0-llk:-KxDTmXRKsV-G1M0-lll:b-bxb4ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/b96f405bf08ed07e643dcd553940b9fb42768e6d/pkg/sql/sqlbase/table.go#L2262)):*
> ```Go
> 	key []byte, validEquivSignatures map[string]int, keySigBuf []byte, keyRestBuf []byte,
> ) (signature []byte, rest []byte, success bool, tableIdx int, rr error) {
> 	signature = signature[:0]
> ```

This now needs to be`= keySigBuf[:0]` and `= keyRestBuf[:0]`. You can also keep the old names and use other names for the return values (`signatureResult restResult`), and never use those names in the code.

Also [nit] but the "key" in these names is not helpful, maybe just `signatureBuf`, `restBuf`.

---

*[pkg/sql/sqlbase/table.go, line 2313 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxDUDvV44WvgtO96EDB:-KxDUDvV44WvgtO96EDC:b-ib4k84) ([raw file](https://github.com/cockroachdb/cockroach/blob/b96f405bf08ed07e643dcd553940b9fb42768e6d/pkg/sql/sqlbase/table.go#L2313)):*
> ```Go
> 		} else {
> 			// The key has been fully decomposed.
> 			break
> ```

[nit] clearer if we just do `if !isSentinel { return ... }`. Then `recentTableIdx` can be scoped to the inside of the loop (`recentTableIdx, found := ..`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Added some solid unit tests, namely for the following cases:
1. `RowFetcher` 1 table fetches with the primary index
2. Fetches from secondary indexes (unique/non-unique with NULL values)
3. A power set of fetches from interleaved tables. Specifically for the interleaved hierarchy:
   ```
   parent1
   parent2
      child1
         grandchild1
           grandgrandchild1
      child2
      grandgrandchild1@ggc1_unique_idx     <-- interleaved on parent2_id (p2)
   parent3
   ```
   there are `2^8 - 1 = 255` test cases to query every non-empty subset of tables.

One takeaway: `MergeSpans` is absolutely necessary during planning to de-duplicate the spans being fed into `MultiRowFetcher` in the case of reading from multiple interleaved tables.

Ready for another look!

RaduBerinde commented:
LGTM  Nice tests!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-:-KxmcYRgInA0O0fne8cq:b-e3ofee)*
<!-- Sent from Reviewable.io -->

richardwu commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/sqlbase/table.go, line 2262 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxDTmXRKsV-G1M0-llk:-KxnYAKmJQKeaJ1zCdRB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b96f405bf08ed07e643dcd553940b9fb42768e6d/pkg/sql/sqlbase/table.go#L2262)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This now needs to be`= keySigBuf[:0]` and `= keyRestBuf[:0]`. You can also keep the old names and use other names for the return values (`signatureResult restResult`), and never use those names in the code.

Also [nit] but the "key" in these names is not helpful, maybe just `signatureBuf`, `restBuf`.
</blockquote></details>

Done.

---

*[pkg/sql/sqlbase/table.go, line 2313 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228#-KxDUDvV44WvgtO96EDB:-KxnYSGxV7A_Y7HnH40t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b96f405bf08ed07e643dcd553940b9fb42768e6d/pkg/sql/sqlbase/table.go#L2313)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] clearer if we just do `if !isSentinel { return ... }`. Then `recentTableIdx` can be scoped to the inside of the loop (`recentTableIdx, found := ..`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19228)*
<!-- Sent from Reviewable.io -->
